src/query/ast/src/ast/format/syntax/query.rs

```diff
@@ -0,0 +1,364 @@
+// Copyright 2022 Datafuse Labs.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use pretty::RcDoc;
+
+use super::common::*;
+use super::expr::expr_doc;
+use crate::ast::Expr;
+use crate::ast::JoinCondition;
+use crate::ast::JoinOperator;
+use crate::ast::OrderByExpr;
+use crate::ast::Query;
+use crate::ast::SelectTarget;
+use crate::ast::SetExpr;
+use crate::ast::SetOperator;
+use crate::ast::TableReference;
+use crate::ast::TimeTravelPoint;
+use crate::ast::With;
+use crate::ast::CTE;
+
+pub(crate) fn query_doc(query: Query) -> RcDoc {
+    with_doc(query.with)
+        .append(body_doc(query.body))
+        .append(order_by_doc(query.order_by))
+        .append(limit_doc(query.limit))
+        .append(offset_doc(query.offset))
+        .append(format_doc(query.format))
+        .group()
+}
+
+fn with_doc(with: Option<With>) -> RcDoc {
```

Comment:
Maybe we can hide `RcDoc` as internal usage, and don't expose to other functions?

Reply:
These functions are private functions and cannot be called from outside.